refactor(compiler-cli): rework type checking for signal forms

crisbeto opened the pull request:
Reworks the way we approach type checking of signal forms to be closer to the behavior at runtime. There are a couple of scenarios that we handle:

1. For native controls, we now produce simplified type checking code that looks as follows:

```
var t1 = null! as number | string; // Type depends on the input `type`.
t1 = someField().value();
```

2. For custom controls we generate bindings to the individual inputs, rather than checking conformance against `FormValueControl`/`FormCheckboxControl`. This is closer to the behavior at runtime and it allows us to handle generic directives properly.

crisbeto committed:
refactor(core): remove unused type

Removes the `ExtractFromControlValue` type since it won't be necessary anymore.

leonsenft commented on packages/compiler-cli/src/ngtsc/typecheck/test/type_check_block_spec.ts:
```diff
@@ -2597,8 +2597,10 @@ describe('type check blocks', () => {
+      expect(block).toContain('var _t1 = null! as string;');
+      expect(block).toContain('_t1 = ((this).f)().value();');
+      expect(block).toContain('var _t2 = null! as i0.Field;');
+      expect(block).toContain('_t2.field = (((this).f));');
```
The PR description states that for native controls, we produce a simplified check (i.e. the value rather than the field), but it looks like we're producing that _in addition_ to the field check. Is that intentional?

leonsenft commented on packages/compiler-cli/test/ngtsc/signal_forms_spec.ts:
```diff
@@ -188,10 +208,41 @@ runInEachFileSystem(() => {
+          @Component({
+            template: \`
+              <custom-control [field]="f" #comp/>
+              {{expectsString(comp.value())}}
```
Just curious, does this have a side-effect of _making_ the generic `T` infer as `string`, or is it just to double check that it _was_ inferred as `string`?

crisbeto committed:
refactor(compiler-cli): rework type checking for signal forms

Reworks the way we approach type checking of signal forms to be closer to the behavior at runtime. There are a couple of scenarios that we handle:

1. For native controls, we now produce simplified type checking code that looks as follows:

```
var t1 = null! as number | string; // Type depends on the input `type`.
t1 = someField().value();
```

2. For custom controls we generate bindings to the individual inputs, rather than checking conformance against `FormValueControl`/`FormCheckboxControl`. This is closer to the behavior at runtime and it allows us to handle generic directives properly.

leonsenft commented on packages/compiler-cli/src/ngtsc/typecheck/test/type_check_block_spec.ts:
```diff
@@ -2685,12 +2693,115 @@ describe('type check blocks', () => {
+        '_t1.value[i1.ɵINPUT_SIGNAL_BRAND_WRITE_TYPE] = i1.ɵunwrapWritableSignal((((((this).f)()).value)));',
+      );
+      expect(block).toContain(
+        '_t1.max[i1.ɵINPUT_SIGNAL_BRAND_WRITE_TYPE] = ((0 as any ? (((((this).f)()).max))!() : undefined));',
```
I'm not sure I completely follow, but I get the exact same error all three ways:

```ts
type Field = () => FieldState;
type Signal<T> = () => T;

interface FieldState {
  get max(): Signal<number | undefined> | undefined;
}

let field: Field = () => ({max: undefined});
let maxInput: string | undefined;

// These all produce identical diagnostics:
maxInput = field().max!();
maxInput = field().max?.();
maxInput = 0 as any ? field().max!() : undefined;
```

I'm I missing some nuance to the issue?

crisbeto replied:
I _think_ there's some way to configure it. It's something that's handled in the same way for all expressions so there's nothing special for signals forms. We just generate a `SafeCall` using the Angular expression AST and then the `tcbExpression` function translates that to TS.

leonsenft replied:
Hmm I guess you're right. It's sort of a convenient coincidence that `max` may be `undefined`, or may be a signal containing `undefined` and using `!` takes advantage of that.

thePunderWoman commented:
This PR was merged into the repository. The changes were merged into the following branches:

- main: f7e58577a43c6e243cfbad8ddec2f7676e25f044
- 21.0.x: be371292f28e2892527ac85e0550426428d20d5f
